pkg/api/v2/types.go

```diff
@@ -0,0 +1,313 @@
+			member := PodGroupMemberSpec{
+				PodMinNumber: memberV1.PodNumber,
+				PodMaxNumber: memberV1.PodNumber,
+				CellsPerPod: PodGroupMemberCellSpec{
+					CellType:   api.CellType(obj.CellType),
+					CellNumber: memberV1.LeafCellNumber,
+				},
+				ContainsCurrentPod: bool(obj.CellNumber == memberV1.LeafCellNumber),
+			}
+			pods = append(pods, member)
+		}
+		obj.PodRootGroup = &PodGroupSpec{
+			Name: objV1.AffinityGroup.Name,
+			Pods: pods,
+		}
+	}
+}
+
+// SetDefaults sets default values for PodSchedulingSpec.
+func (obj *PodSchedulingSpec) SetDefaults(pod *core.Pod) {
+	if obj.PodRootGroup == nil {
+		obj.PodRootGroup = &PodGroupSpec{
+			Name: fmt.Sprintf("%v/%v", pod.Namespace, pod.Name),
+			Pods: []PodGroupMemberSpec{{
+				PodMinNumber: 1,
+				PodMaxNumber: 1,
+				CellsPerPod: PodGroupMemberCellSpec{
+					CellType:   api.CellType(obj.CellType),
+					CellNumber: obj.CellNumber,
+				},
+				ContainsCurrentPod: true,
+			}},
+		}
+	}
+}
+
+// Validate checks whether PodSchedulingSpec is ok.
+func (obj *PodSchedulingSpec) Validate() (msg string, ok bool) {
+	if obj.VirtualCluster == "" {
+		return "VirtualCluster is empty", false
+	}
+	if obj.Priority < api.OpportunisticPriority {
+		return fmt.Sprintf("Priority is less than %v", api.OpportunisticPriority), false
+	}
+	if obj.Priority > api.MaxGuaranteedPriority {
+		return fmt.Sprintf("Priority is greater than %v", api.MaxGuaranteedPriority), false
+	}
+	if obj.CellNumber <= 0 {
+		return "CellNumber is non-positive", false
+	}
+	if obj.PodRootGroup.Name == "" {
+		return "PodRootGroup.Name is empty", false
+	}
+
+	isPodInGroup := false
+	queue := []*PodGroupSpec{obj.PodRootGroup}
+	for len(queue) > 0 {
+		newQueue := []*PodGroupSpec{}
+		for _, podGroup := range queue {
+			for _, pod := range podGroup.Pods {
+				if pod.PodMinNumber <= 0 {
+					return "PodGroup.Pods have non-positive PodMinNumber", false
+				}
+				if pod.PodMaxNumber <= 0 {
+					return "PodGroup.Pods have non-positive PodMaxNumber", false
+				}
+				if pod.CellsPerPod.CellNumber <= 0 {
+					return "PodGroup.Pods have non-positive CellsPerPod.CellNumber", false
+				}
+				if pod.ContainsCurrentPod == true {
+					if isPodInGroup == false {
+						isPodInGroup = true
+					} else {
+						return "PodGroup.Pods have multiple ContainsCurrentPod", false
+					}
+				}
+			}
+			newQueue = append(newQueue, podGroup.ChildGroups...)
+		}
+		queue = newQueue
+	}
+	if !isPodInGroup {
+		return "PodGroup.Pods does not contain current Pod", false
+	}
+	return "", true
+}
+
+type PodBindingInfo struct {
+	Node                    string               `yaml:"node"`              // node to bind
+	LeafCellIsolation       []int32              `yaml:"leafCellIsolation"` // leaf cells to bind
+	CellChain               string               `yaml:"cellChain"`         // cell chain selected
+	PodRootGroupBindingInfo *PodGroupBindingInfo `yaml:"PodRootGroupBindingInfo"`
+}
+
+type PodGroupBindingInfo struct {
```

Comment:
>PodGroupBindingInfo [](http://example.com/codeflow?start=5&length=19)

Suggest to still name as PodGroupBindInfo, see comment for PodBindingInfo

Reply:
renamed